pilot/pkg/serviceregistry/kube/mcs/discoverability.go

```diff
@@ -0,0 +1,310 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mcs
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+
+	"istio.io/istio/pilot/pkg/model"
+	kubesr "istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/config/host"
+	"istio.io/istio/pkg/config/schema/gvk"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/queue"
+	"istio.io/pkg/log"
+	"istio.io/pkg/monitoring"
+)
+
+const (
+	discoverabilityLogPrefix = "MCS DiscoverabilityController:"
+)
+
+var (
+	empty = struct{}{}
+
+	eventTag   = monitoring.MustCreateLabel("event")
+	clusterTag = monitoring.MustCreateLabel("cluster")
+
+	serviceExportEvents = monitoring.NewSum(
+		"service_export_events",
+		"MCS ServiceExport events fromm all k8s clusters.",
+		monitoring.WithLabels(eventTag, clusterTag),
+	)
+)
+
+var _ model.ServiceDiscoverabilityPolicy = &DiscoverabilityController{}
+
+func init() {
+	monitoring.MustRegister(serviceExportEvents)
+}
+
+func incrementEvent(event string, cluster kube.ClusterName) {
+	serviceExportEvents.With(eventTag.Value(event), clusterTag.Value(string(cluster))).Increment()
+}
+
+type DiscoverabilityConfig struct {
+	Clusters     *kube.ClusterSet
+	XDSUpdater   model.XDSUpdater
+	DomainSuffix string
+}
+
+// NewDiscoverabilityController creates a new DiscoverabilityController that observes the given ClusterSet.
+func NewDiscoverabilityController(cfg DiscoverabilityConfig) *DiscoverabilityController {
+	c := &DiscoverabilityController{
+		clusters:     cfg.Clusters,
+		xdsUpdater:   cfg.XDSUpdater,
+		domainSuffix: cfg.DomainSuffix,
+		exports:      make(map[kube.ClusterName]*exportedServices),
```

Comment:
nit: do we need pointer to exportedServices since everything inside exportedServices is already pointers?